File: Neetcode/1DDP/coinchange.py
# https://neetcode.io/problems/coin-change
# https://leetcode.com/problems/coin-change/

from typing import List


class olution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        track = [99999] * (amount + 1)
        track[0] = 0

        for i in range(amount + 1):
            for c in coins:
                if i + c < amount + 1:
                    track[i + c] = min(track[i] + 1, track[i + c])
        return track[amount] if track[amount] != 99999 else -1

File: Neetcode/1DDP/coinchange.cpp
#include <vector>
using namespace std;

class Solution {
 public:
  int coinChange(vector<int>& coins, int amount) {
    vector<int> track(amount + 1, 99999);
    track[0] = 0;
    for (int i = 1; i < amount + 1; i++) {
      for (int coin : coins) {
        if (coin <= i) {
          track[i] = min(track[i - coin] + 1, track[i]);
        }
      }
    }
    return (track[amount] != 99999) ? track[amount] : -1;
  }
};